Feat/redis

shay123g opened the pull request:
Closes #29 
basic Redis DB functionality
*Store
*Get

without configuration impl nor tests
@SnirShechter need ur help with config implementation and test writing

SnirShechter commented:
* Remember that this Redis instance might be used by others as well. Add prefixes to the record keys, e.g 'miniurl:url:<ID>'.
* You might not need === 'OK' - It would probably throw if it fails (check it)
* The Redis helper can just be replaced with the regular get function in the driver (just use it for all other places)
```typescript
		get(id: string, options: { withInfo: boolean }): Promise<StoredUrl | UrlWithInformation> {
			const storedUrl = JSON.parse(await this.redis.get(id))
			
			if(!withInfo) return {
				id: storedUrl.id,
				url: storedUrl.url,
				createdAt: storedUrl.createdAt,
				updatedAt: storedUrl.updatedAt,
			}
			
			return storedUrl
		}
```
* Passing the config
	* The raw config is straight from the environment variables so it should stay all strings (I changed it)
	* Go to src/config/validate.ts and make sure you validated everything you need (and can validate). be very strict about it, e.g if host is not defined, throw an error.
	* On src/services/storage/drivers/redis/index.ts, remove all uses of the config imported from src/config. If you need that config, add it to the driverConfig instead.
	* The src/config.normalize.ts is the place to change the config how you want it. For example, the port is a string in RawConfig. If you need it as a number, this is the place to change it. If you need the config.lifetimeMs, add it here and so on.
	* On src/config.normalize.ts, change the storage.driverConfig to include the Redis config if driverName === 'Redis'. The config here is the config that the Redis driver will receive.
	* On src/services/storage/drivers/redis/types.ts, edit the interface to contain that config. If that config is the IORedis config, just extend it (see the equivalent Relational config).
	* On src/services/storage/drivers/redis/index.ts in the constructor, the _config argument is the config for you. Create the new IORedis() inside the constructor function (this._client = new IORedis(config.port, config.host) etc. or similar) and pass the config to it.
* Logs - there were too many logs. I removed most of them. If there are specific cases needed to log, re-add them.
* Errors - OperationFailed error is inside generalError.ts. Move it to its own file or move all errors to the same file (I think I prefer the latter, since every error is just 1 line)

shay123g commented on package.json:
```diff
@@ -40,23 +40,26 @@
 		"pg": "^8.5.1",
 		"pino": "^6.11.1",
 		"snake-case": "^3.0.4",
+		"tedis": "^0.1.12",
```
done

shay123g commented on package.json:
```diff
@@ -40,23 +40,26 @@
 	"devDependencies": {
 		"@istanbuljs/nyc-config-typescript": "^1.0.1",
+		"@types/ioredis": "^4.22.0",
+		"@types/pino": "^6.3.6",
```
done

SnirShechter replied:
Seems like it didn't change

shay123g commented:
> * Remember that this Redis instance might be used by others as well. Add prefixes to the record keys, e.g 'miniurl:url:'.
> * You might not need === 'OK' - It would probably throw if it fails (check it)
> * The Redis helper can just be replaced with the regular get function in the driver (just use it for all other places)
> 
> ```ts
> 		get(id: string, options: { withInfo: boolean }): Promise<StoredUrl | UrlWithInformation> {
> 			const storedUrl = JSON.parse(await this.redis.get(id))
> 			
> 			if(!withInfo) return {
> 				id: storedUrl.id,
> 				url: storedUrl.url,
> 				createdAt: storedUrl.createdAt,
> 				updatedAt: storedUrl.updatedAt,
> 			}
> 			
> 			return storedUrl
> 		}
> ```
> 
> * Passing the config
>   
>   * The raw config is straight from the environment variables so it should stay all strings (I changed it)
>   * Go to src/config/validate.ts and make sure you validated everything you need (and can validate). be very strict about it, e.g if host is not defined, throw an error.
>   * On src/services/storage/drivers/redis/index.ts, remove all uses of the config imported from src/config. If you need that config, add it to the driverConfig instead.
>   * The src/config.normalize.ts is the place to change the config how you want it. For example, the port is a string in RawConfig. If you need it as a number, this is the place to change it. If you need the config.lifetimeMs, add it here and so on.
>   * On src/config.normalize.ts, change the storage.driverConfig to include the Redis config if driverName === 'Redis'. The config here is the config that the Redis driver will receive.
>   * On src/services/storage/drivers/redis/types.ts, edit the interface to contain that config. If that config is the IORedis config, just extend it (see the equivalent Relational config).
>   * On src/services/storage/drivers/redis/index.ts in the constructor, the _config argument is the config for you. Create the new IORedis() inside the constructor function (this._client = new IORedis(config.port, config.host) etc. or similar) and pass the config to it.
> * Logs - there were too many logs. I removed most of them. If there are specific cases needed to log, re-add them.
> * Errors - OperationFailed error is inside generalError.ts. Move it to its own file or move all errors to the same file (I think I prefer the latter, since every error is just 1 line)

- moved all the errors to one file

SnirShechter commented on .env.development:
```diff
@@ -5,8 +5,8 @@ LOG_LEVEL=debug
-## InMemory/Relational
-STORAGE_DRIVER=InMemory
+## InMemory/Relational/Redis
+STORAGE_DRIVER=Redis
```
Leave `InMemory` as the default

SnirShechter commented on .env.development:
```diff
@@ -15,3 +15,11 @@ STORAGE_DRIVER=InMemory
+ REDIS_STORAGE_PORT = 6379
+ REDIS_STORAGE_HOST = localhost
+ #REDIS_STORAGE_USERNAME = redis
+#REDIS_STORAGE_PASSWORD = redis
```
Indent these lines and add a guide to run Redis in the contribution.md

SnirShechter commented on src/services/storage/drivers/redis/types.ts:
```diff
@@ -0,0 +1,6 @@
+import { config } from '../../../../config/index.js'
```
This file is strictly for types. Don't import the config here.

If you need the urlLifeTime from the config, just copy it in the config module.

shay123g commented:
**notes to self for today's evening**
1."@types/ioredis": "^4.22.0",
  "@types/pino": "^6.3.6",
take out of devDependecies
2.env.development - default should be STORAGE_DRIVER=InMemory
3. types.ts file only for types - do not import anything else like config
4. prettify imports: combine imported objects in same line like import {RedisStorageConfig, StorageDriverName} from...
5. generalError in erros file should  be **G**eneralError
6. remove RedisDriverConfig section from Config object. it should be converted in normalize.ts and inserted into storageDriverConfiglike in Relational...
7. to be able to use  RedisStorageConfig extends IORedis.Options, remove the 'extends...' from RelationalStorageConfig
8. delete tests (or keep in side branch) - moving to Jest
9. **most important: seperate to differents PR**
@SnirShechter  I will delete this PR and make new ones, more neat.. pls add here any point i forget from our zoom meting
